### PR TITLE
AP_Arming: force longer disarm logging for all nasty disarm reasons

### PR DESCRIPTION
**Pull request by IamPete1**
https://github.com/ArduPilot/ardupilot/pull/8836 https://github.com/ArduPilot/ardupilot/issues/8384

Love that we have both `TERRRAINFAILSAFE` and `TERRAINFAILSAFE` cant blame me for that one ;)

### CODE REVIEW

**Comment by peterbarker** [comment]
We should already persist logging for 15 seconds in any circumstance.

Perhaps we should just remove the force-disarm-logging entirely?  It predates the 15-second thing.

**Comment by IamPete1** [comment]
15 seconds is not enough to log all of a decent under parachute from any largish altitude. I don't see that harm in just carrying on logging in such cases.

File: libraries/AP_Arming/AP_Arming.cpp [hunk 1]
```diff
@@ -1453,6 +1453,27 @@ void AP_Arming::Log_Write_Disarm(const AP_Arming::Method method)
     };
     AP::logger().WriteCriticalBlock(&pkt, sizeof(pkt));
     AP::logger().Write_Event(LogEvent::DISARMED);
+
+    if (method == AP_Arming::Method::TERMINATION ||
```
Comment (peterbarker) [inline libraries/AP_Arming/AP_Arming.cpp:1457]: If you use a switch in here instead then when someone adds a new method they are forced to evaluate whether it should be added here.

File: libraries/AP_Arming/AP_Arming.cpp [hunk 2]
```diff
@@ -1453,6 +1453,27 @@ void AP_Arming::Log_Write_Disarm(const AP_Arming::Method method)
     };
     AP::logger().WriteCriticalBlock(&pkt, sizeof(pkt));
     AP::logger().Write_Event(LogEvent::DISARMED);
+
```
Comment (peterbarker) [inline libraries/AP_Arming/AP_Arming.cpp:1456]: This shouldn't be in `Log_Write_Disarm`.  There's only one caller to this function - it should go next to that call.  We *may* want a new `check_forced_logging()` method or something.

**Review by peterbarker — APPROVED** [review]
LGTM.

**Comment by IamPete1** [comment]
moved `Method::BADFLOWOFCONTROL` to force logging.

**Comment by tridge** [comment]
I'd prefer a param for how long to log after disarm in seconds

**Comment by IamPete1** [comment]
https://github.com/ArduPilot/ardupilot/issues/10148

**Review by tridge — APPROVED** [review]

**Pull request merged**